custom_components/dreo/climate.py

```diff
@@ -0,0 +1,350 @@
+
+# Suppress warnings about unused function arguments
+# pylint: disable=W0613
+from __future__ import annotations
+
+from typing import Any
+from dataclasses import dataclass
+import logging
+
+from .haimports import * # pylint: disable=W0401,W0614
+from .basedevice import DreoBaseDeviceHA
+from .pydreo import PyDreo
+from .pydreo.pydreobasedevice import PyDreoBaseDevice
+from .pydreo import (
+    PyDreoHeater,
+    HEATER_MODE_OFF,
+    HEATER_MODE_COOLAIR,
+    HEATER_MODE_HOTAIR,
+    HEATER_MODE_ECO,
+    MODE_LEVEL_MAP,
+    LEVEL_MODE_MAP,
+    HEAT_RANGE,
+    ECOLEVEL_RANGE
+)
+
+from homeassistant.components.climate import (
+    ATTR_TEMPERATURE,
+    FAN_ON,
+    FAN_OFF,
+    PRESET_ECO,
+    SWING_OFF,
+    SWING_ON,
+    ClimateEntity,
+    ClimateEntityFeature,
+    HVACAction,
+    HVACMode,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    ATTR_ENTITY_ID,
+    ATTR_TEMPERATURE,
+    PRECISION_HALVES,
+    PRECISION_TENTHS,
+    PRECISION_WHOLE,
+    STATE_OFF,
+    STATE_ON,
+    UnitOfTemperature,
+)
+from homeassistant.core import HomeAssistant, ServiceCall
+from homeassistant.helpers import entity_platform
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.temperature import display_temp as show_temp
+
+from .const import (
+    LOGGER,
+    DOMAIN,
+    DREO_MANAGER
+)
+
+HVAC_MODE_MAP = { 
+    HVACMode.OFF: HEATER_MODE_OFF,
+    HVACMode.FAN_ONLY: HEATER_MODE_COOLAIR,
+    HVACMode.AUTO: HEATER_MODE_ECO,
+    HVACMode.HEAT: HEATER_MODE_HOTAIR,
+}
+
+HEATER_MODE_MAP = { 
+    HEATER_MODE_OFF : HVACMode.OFF,
+    HEATER_MODE_COOLAIR : HVACMode.FAN_ONLY,
+    HEATER_MODE_HOTAIR : HVACMode.HEAT,
+    HEATER_MODE_ECO : HVACMode.AUTO
+}
+
+SWING_MODES = [
+    SWING_OFF, 
+    SWING_ON
+]
+
+_LOGGER = logging.getLogger(LOGGER)
+
+def add_device_entries(devices) -> []:
+    heater_ha_collection = []
+    
+    for de in devices:
+        _LOGGER.debug("Adding heater for %s --> de is %s", de.name, de)
+        dt = DreoHeaterHA(de)
+        heater_ha_collection.append(dt)
+    
+    return heater_ha_collection
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the Dreo Heater platform."""
+    _LOGGER.info("Starting Dreo Heater Platform")
+    _LOGGER.debug("Dreo Heater:async_setup_platform")
+
+    manager : PyDreo = hass.data[DOMAIN][DREO_MANAGER]
+
+    async_add_entities(add_device_entries(manager.heaters))
+
+    platform = entity_platform.async_get_current_platform()
+
+
+
+
+# Implementation of the heater
+class DreoHeaterHA(DreoBaseDeviceHA, ClimateEntity):
+    """Representation of a Dreo heater as a climate entity."""
+    _attr_precision = PRECISION_WHOLE
+    _attr_temperature_unit = UnitOfTemperature.FAHRENHEIT # TODO
+    _attr_target_temperature = None
+    _attr_current_temperature = None
+    _attr_fan_mode = None
```

Comment:
Is it worth refactoring Fan/Heater so they share more code?

Reply:
At some point yes, but initially I wanted to keep them separate and not touch the Fan code too much because I didn't want to break anything. Not having a fan myself, I thought it best if I keep it separate for the moment.